### PR TITLE
Fix issue #108

### PR DESCRIPTION
**Pull request by nerosketch**
fix SystemError "returned NULL without setting an error" in python3.7.
i was run it with little fixes under python3.7. I hope it will help anyone.

### CODE REVIEW

File: easysnmp/interface.c [hunk 1]
```diff
@@ -1517,7 +1517,14 @@ static PyObject *create_session_capsule(SnmpSession *session)
     ctx->handle = handle;
     ctx->invalid_oids = (bitarray *) ctx->invalid_oids_buf;
     bitarray_buf_init(ctx->invalid_oids, sizeof(ctx->invalid_oids_buf));
-    return (capsule);
+    if(capsule == NULL)
+    {
+        Py_RETURN_NONE;
```
Comment (ngie-eign) [inline easysnmp/interface.c]: Could you please simplify/fix the style like so?
```
if (capsule == NULL)
{
    Py_RETURN_NONE;
}
return capsule;
```

**Comment by ngie-eign** [comment]
@kamakazikamikaze -san: could you please integrate this change into `easysnmp` after the style fix is made?

File: easysnmp/interface.c [hunk 2]
```diff
@@ -2226,9 +2233,9 @@ static PyObject *netsnmp_get(PyObject *self, PyObject *args)
 
     if (error)
```
Comment (ngie-eign) [inline easysnmp/interface.c]: Actually, the `if (error) {...}` block can be deleted, since the two cases both converge to `Py_RETURN_NONE;`, so:
```
if (error)
{
    Py_RETURN_NONE;
}
Py_RETURN_NONE;
```
should just be:
```
Py_RETURN_NONE;
```

Going a step further, it looks like all of the cleanup is properly handling this case, so I guess the use of `error` could also be removed.

This basic observation can be applied elsewhere.

**Review by ngie-eign — CHANGES_REQUESTED** [review]
Actually, looking at the review and the docs some more (because I had forgotten these nuances), this is the wrong solution to the problem: https://stackoverflow.com/a/8450880 . The real problem is that an exception isn't being properly raised on error, ergo, it's throwing a programming error of sorts. Do not merge this change.

**Comment by nerosketch** [comment]
> I'd reword this to something like: `` // Don't convert if `attr` is already in bytes. ``

Thank you for your code review. You are right, I wasn't familiar with the subject. Please check it again. Now more errors return correctly, but I not tested all functions, only netsnmp_get and netsnmp_walk. In few days I'm planned to test it all. P.S. Sorry for my english.

File: easysnmp/interface.c [hunk 3]
```diff
@@ -444,7 +444,7 @@ static int __snprint_value(char *buf, size_t buf_len,
             case ASN_TIMETICKS:
             case ASN_UINTEGER:
                 snprintf(buf, buf_len, "%lu",
-                         (unsigned long) *var->val.integer);
+                         (ulong) *var->val.integer);
```
Comment (ngie-eign) [inline easysnmp/interface.c]: This is unnecessary churn. Could you please change the cast back to `unsigned long` here?

File: easysnmp/interface.c [hunk 4]
```diff
@@ -1517,7 +1546,9 @@ static PyObject *create_session_capsule(SnmpSession *session)
     ctx->handle = handle;
     ctx->invalid_oids = (bitarray *) ctx->invalid_oids_buf;
     bitarray_buf_init(ctx->invalid_oids, sizeof(ctx->invalid_oids_buf));
-    return (capsule);
+
+    return capsule;
```
Comment (ngie-eign) [inline easysnmp/interface.c:1708]: I generally discourage unnecessary churn like this, but this seems to be making the style of the line of code more in line with the general style.

It might make more sense to submit this as a separate review though, since the line wasn't affected by your overall proposed changeset.

File: easysnmp/interface.c [hunk 5]
```diff
@@ -1393,7 +1422,7 @@ static long long py_netsnmp_attr_long(PyObject *obj, char *attr_name)
     return val;
 }
 
-static void *py_netsnmp_attr_void_ptr(PyObject *obj, char *attr_name)
+/*static void *py_netsnmp_attr_void_ptr(PyObject *obj, char *attr_name)
```
Comment (ngie-eign) [inline easysnmp/interface.c]: It's better to use `#if 0` -> `#endif` for code that's going to be disabled over the short-term. It's better to just remove it, however, in a separate commit if it's unused (it can be restored later as need be from git).

Using multi-line comments is bad because (depending on the compiler), it can flag the issue and fail builds with `-Werror` enabled.

File: easysnmp/interface.c [hunk 6]
```diff
@@ -1352,12 +1352,31 @@ static int py_netsnmp_attr_string(PyObject *obj, char *attr_name, char **val,
...
-            // retrieve its value and length
-            PyObject *attr_bytes = PyUnicode_AsEncodedString(attr, "latin-1",
-                                                             "surrogateescape");
+
+            PyObject *attr_bytes;
+            if(PyBytes_CheckExact(attr))
+            {
+                // Don't convert if `attr` is already in bytes.
+                attr_bytes = attr;
+            }else if(PyUnicode_CheckExact(attr))
```
Comment (ngie-eign) [inline easysnmp/interface.c]: style bug:
```
}else if(PyUnicode_CheckExact(attr))
{
```
should be:
```
}
else if (PyUnicode_CheckExact(attr))
{
```

File: easysnmp/interface.c [hunk 7]
```diff
@@ -1352,12 +1352,31 @@ static int py_netsnmp_attr_string(PyObject *obj, char *attr_name, char **val,
...
+            {
+                // Encode the provided attribute using latin-1 into bytes and
+                // retrieve its value and length
+                attr_bytes = PyUnicode_AsEncodedString(
+                        attr, "latin-1", "surrogateescape");
+            }else
```
Comment (ngie-eign) [inline easysnmp/interface.c]: style bug:
```
}else
{
```
should be
```
}
else
{
```

File: easysnmp/interface.c [hunk 8]
```diff
@@ -1352,12 +1352,31 @@ static int py_netsnmp_attr_string(PyObject *obj, char *attr_name, char **val,
             int retval;
 
 #if PY_MAJOR_VERSION >= 3
-            // Encode the provided attribute using latin-1 into bytes and
-            // retrieve its value and length
-            PyObject *attr_bytes = PyUnicode_AsEncodedString(attr, "latin-1",
-                                                             "surrogateescape");
+
+            PyObject *attr_bytes;
+            if(PyBytes_CheckExact(attr))
```
Comment (ngie-eign) [inline easysnmp/interface.c:1436]: style bug: `if(PyBytes_CheckExact(attr))` should be `if (PyBytes_CheckExact(attr))`

File: easysnmp/interface.c [hunk 9]
```diff
@@ -1370,7 +1389,17 @@ static int py_netsnmp_attr_string(PyObject *obj, char *attr_name, char **val,
 
             Py_DECREF(attr);
             return retval;
+        }else
```
Comment (ngie-eign) [inline easysnmp/interface.c]: style bug:
```
}else
{
```
should be
```
}
else
{
```

File: easysnmp/interface.c [hunk 10]
```diff
@@ -1370,7 +1389,17 @@ static int py_netsnmp_attr_string(PyObject *obj, char *attr_name, char **val,
...
+        {
+            PyErr_Format(PyExc_AttributeError,
+                         "Failed to get '%s' attribute",
+                         attr_name);
         }
+    }else
```
Comment (ngie-eign) [inline easysnmp/interface.c]: style bug:
```
}else
{
```
should be
```
}
else
{
```

File: easysnmp/interface.c [hunk 11]
```diff
@@ -1142,12 +1142,12 @@ static int __send_sync_pdu(netsnmp_session *ss, netsnmp_pdu *pdu,
                            bitarray *invalid_oids)
 {
     int status = 0;
-    long command = pdu->command;
+    int command = pdu->command;
     char *tmp_err_str;
     size_t retry_num = 0;
 
     /* Note: SNMP uses 1-based indexing with OIDs, so 0 is unused */
-    unsigned long last_errindex = 0;
```
Comment (ngie-eign) [inline easysnmp/interface.c:1203]: This is unnecessary churn. Could you please change the cast back to `unsigned long` here?

File: easysnmp/interface.c [hunk 12]
```diff
@@ -146,8 +146,8 @@ static int __get_type_str(int type, char *str, int log_error);
 static int __get_label_iid(char *name, char **last_label, char **iid,
                            int flag);
 static struct tree *__tag2oid(char *tag, char *iid, oid *oid_arr,
-                              int *oid_arr_len, int *type, int best_guess);
-static int __concat_oid_str(oid *doid_arr, int *doid_arr_len, char *soid_str);
+                              size_t *oid_arr_len, int *type, int best_guess);
+static int __concat_oid_str(oid *doid_arr, size_t *doid_arr_len, char *soid_str);
```
Comment (ngie-eign) [inline easysnmp/interface.c]: If these types are being converted from `int` to `size_t`, there are other parameters in other functions that need to be changed. Example:
```
static int __add_var_val_str(netsnmp_pdu *pdu, oid *name, int name_length,
                             char *val, int len, int type);
```
should be:
```
static int __add_var_val_str(netsnmp_pdu *pdu, oid *name, size_t name_length,
                             char *val, size_t len, int type);
```
I feel like this is going to be like peeling an onion. It would be better to have these changes be merged separately to keep the commit history for the lines clean.

File: easysnmp/interface.c [hunk 13]
```diff
@@ -1142,12 +1142,12 @@ static int __send_sync_pdu(netsnmp_session *ss, netsnmp_pdu *pdu,
                            bitarray *invalid_oids)
 {
     int status = 0;
-    long command = pdu->command;
```
Comment (ngie-eign) [inline easysnmp/interface.c]: Revert this change. It doesn't align with the definition of `netsnmp_pdu`: http://net-snmp.sourceforge.net/dev/agent/structsnmp__pdu.html .

File: easysnmp/interface.c [hunk 14]
```diff
@@ -3788,6 +3823,10 @@ static PyObject *netsnmp_set(PyObject *self, PyObject *args)
     {
         if (!PyArg_ParseTuple(args, "OO", &session, &varlist))
         {
+            error = 1;
+            PyErr_SetString(PyExc_ValueError,
```
Comment (ngie-eign) [inline easysnmp/interface.c]: This can be condensed to a single line:
```
            PyErr_SetString(PyExc_ValueError,
                    "missing 2 required arguments: 'session' and 'varlist'");
```

**Review by ngie-eign — CHANGES_REQUESTED** [review]
Thank you very much for working on this!

The overall change seems to be on the right path; the amount of churn just needs to be reduced and some of the style bugs need to be fixed.

**Comment by ngie-eign** [comment]
@nerosketch: Ping about my CR comments?

File: easysnmp/interface.c (hunk 3 above)
Comment (nerosketch) [reply easysnmp/interface.c]: done, thanks

File: easysnmp/interface.c (hunk 4 above)
Comment (nerosketch) [reply easysnmp/interface.c:1708]: done, thanks

File: easysnmp/interface.c (hunk 5 above)
Comment (nerosketch) [reply easysnmp/interface.c]: done, thanks

File: easysnmp/interface.c (hunk 6 above)
Comment (nerosketch) [reply easysnmp/interface.c]: done, thanks

File: easysnmp/interface.c (hunk 14 above)
Comment (nerosketch) [reply easysnmp/interface.c]: ok

File: easysnmp/interface.c (hunk 7 above)
Comment (nerosketch) [reply easysnmp/interface.c]: yes, you're right. done, thanks

File: easysnmp/interface.c (hunk 10 above)
Comment (nerosketch) [reply easysnmp/interface.c]: did

File: easysnmp/interface.c (hunk 13 above)
Comment (nerosketch) [reply easysnmp/interface.c]: I've seen **int** type for **command** field in source file. In orig code long type now. I don't quite understand. Need to go back to long again?

File: easysnmp/interface.c (hunk 11 above)
Comment (nerosketch) [reply easysnmp/interface.c:1203]: reverted

File: easysnmp/interface.c (hunk 12 above)
Comment (nerosketch) [reply easysnmp/interface.c]: I think this is too hard for me. If it doesn't affect the functionality too much then i would prefer leave it as it is. I looked up all type changes which i made, and all dependent code that it affects, and i hope it all has been resolved. Need to add more unit tests for check it more carefully. I i'll try this

Comment (ngie-eign) [reply easysnmp/interface.c]: > I think this is too hard for me. If it doesn't affect the functionality too much then i would prefer leave it as it is. I looked up all type changes which i made, and all dependent code that it affects, and i hope it all has been resolved. Need to add more unit tests for check it more carefully. I i'll try this

Ok. This should probably be a series of follow up commits.

File: easysnmp/interface.c (hunk 13 above)
Comment (ngie-eign) [reply easysnmp/interface.c]: *sigh* ok, this should be resolved in a set of follow up commits. My concern is that there are going to be compiler errors triggered by `-Woverflow`, etc, but the problems will need to be resolved in a single shot.

Comment (nerosketch) [reply easysnmp/interface.c]: > _sigh_ ok, this should be resolved in a set of follow up commits. My concern is that there are going to be compiler errors triggered by `-Woverflow`, etc, but the problems will need to be resolved in a single shot.

I'd rather agree. But I don't fully understand this part. I'm sorry, please, if you have time to explain this section to me i'll be grateful. Why do we need type "long" if in the source header "command" field have "int" type? Will there be an overflow when we copying from long to int in netsnmp_pdu structure?

I mean, if i have:
long command = pdu->command;

pdu->command have "int" type, but in "command" var I have "long" type.
If i will use two different type - it's seems like overflow can happen, instead of if I will use the same type for both - than there should be no overflow.

Is it right?
Or maybe you can advise me some source where i can read and learn about it and find my answers. Thanks.

Comment (nerosketch) [reply easysnmp/interface.c]: It variable uses later in "snmp_fix_pdu" function in second parameter. And second parameter accepts int too

Comment (ngie-eign) [reply easysnmp/interface.c]: > Will there be an overflow when we copying from long to int in netsnmp_pdu structure?

There could be if we're not careful (monotonic counters are notorious for causing problems).

However, if the underlying type for some of the APIs is already int, then the issue will need to be addressed in the library as another series of commits to prevent overflow from happening when an int overflows, as (2^32) / 2-1 -> 2,147,483,647, whereas (2^64)/2 - 1 -> 9.223372036854776e+18.

Comment (nerosketch) [reply easysnmp/interface.c]: Thank you, i did

**Comment by bernhardschmidt** [comment]
I was hit by #108 and this patch seems to work fine. I built a Debian package for it. The apply procedure was a bit cumbersome because the Debian build process does not deal well with concatenated patches (exporting this PR into a single .patch file) and there are a few patch chunks that have already been in v0.2.5 (again something the Debian build cannot cope with), but overall it was sane.

Since the python net-snmp bindings have been dropped from Debian I'm looking for an alternative. If we can get this reviewed/merged I'd be willing to make an official Debian package from this code. Might help with finding a new maintainer.

**Comment by uedvt359** [comment]
This patch does (afaiu) not address the root cause of issue 108, and instead adds `PyErr_SetString()` calls at various locations to work around the actual problem (that the exception is swallowed in `__py_netsnmp_update_session_errors()`). I am proposing https://github.com/kamakazikamikaze/easysnmp/pull/129 as a replacement to this pull request.

**Comment by kamakazikamikaze** [comment]
I'm inclined to agree with @uedvt359 that this did not necessarily fix the underlying issues. However, there are some pieces of code here that may fix up certain warnings or undesired behavior. I may cherry-pick some of the commits to add to the repo.

I apologize for my absence in development of the repo these past few years but I hope that anyone who has been waiting for these fixes to be merged will finally get non-erratic behavior from the package.
